### PR TITLE
feature : Implement node2vec algorithm

### PR DESCRIPTION
**Pull request by vinayhpandya**
- Added mlx_cluster for fast random walk implementation
- Implement node2vec algorithm

## Proposed changes

Implements node2vec algorithm from the paper `node2vec: Scalable Feature Learning for Networks` using custom random walk function from mlx_cluster

## Checklist

Put an `x` in the boxes that apply.
- [x ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have updated the necessary documentation

## TODO
- Implement metal kernels for biased random walks

### CODE REVIEW

File: mlx_graphs/nn/models/node2vec.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+            positive samples. This parameter increases the effective sampling
+            rate by reusing samples across different source nodes.
+        num_nodes (int): Number of nodes in a graph
+        walks_per_node (int, optional): The number of walks to sample for each
+            node. (default: :obj:`1`)
+        p (float, optional): Likelihood of immediately revisiting a node in the
+            walk. (default: :obj:`1`)
+        q (float, optional): Control parameter to interpolate between
+            breadth-first strategy and depth-first strategy (default: :obj:`1`)
+        num_negative_samples (int, optional): The number of negative samples to
```
Comment (francescofarina) [inline mlx_graphs/nn/models/node2vec.py:37]: you can safely remove the types from here as they're inferred from the __init__ signature

**Comment by francescofarina** [comment]
@vinayhpandya looks like tests are failing on python 3.10 and 3.12 - did you test `mlx_cluster` on those?

**Comment by vinayhpandya** [comment]
@francescofarina I am currently testing on those, it seems to be an issue with `fmt` library while compiling mlx, I will try to fix that today

**Comment by vinayhpandya** [comment]
> @vinayhpandya looks like tests are failing on python 3.10 and 3.12 - did you test `mlx_cluster` on those?

It seems to be an issue in mlx compilation itself which should be fixed in next release

**Comment by TristanBilot** [comment]
hi @vinayhpandya, thanks for adding node2vec to the lib.
Could you add an example on how using it? Like a training loop where you use both the forward and loss functions. thx!

**Comment by vinayhpandya** [comment]
> @vinayhpandya looks like tests are failing on python 3.10 and 3.12 - did you test `mlx_cluster` on those?

Yes, I created a temporary notebook for training and will publish a PR for it by early next week

**Comment by vinayhpandya** [comment]
@TristanBilot I will raise the PR for a simple notebook by early next week. As a new version of mlx has been pushed out, I need to fix some changes in my library and push it to pypi

**Comment by francescofarina** [comment]
@vinayhpandya can you remove types in the doctoring here https://github.com/mlx-graphs/mlx-graphs/pull/170#discussion_r1766946622 ?

I'm also not sure whether `nn/models/node2vec.py` is the best place to put the node2vec algorithm. I think `nn` should be used for architectural building blocks and I'm not sure node2vec qualifies for that - @TristanBilot what do you think?

**Comment by vinayhpandya** [comment]
@francescofarina I think I can move the it to a separate folder if we don't want to have it in `nn` folder

**Comment by francescofarina** [comment]
> @francescofarina I think I can move the it to a separate folder if we don't want to have it in `nn` folder

Yes, perhaps we can have an `algorithms` module or something similar.

**Comment by francescofarina** [comment]
@vinayhpandya I moved it to a standalone `algorithms` module instead of having it as a submodule of `nn` and added it to the docs.

**Review by francescofarina — APPROVED** [review]
Looks good to me - I'll work on a new release.

**Comment by TristanBilot** [comment]
Seems good to me! Thanks @vinayhpandya
